crates/containerd-shim-wasm/src/sandbox/cgroups.rs

```diff
@@ -0,0 +1,674 @@
+                    let path = self.get_file(&format!("io.max"));
+                    ensure_write_file(
+                        path,
+                        &format!(
+                            "{}:{} wbps={}",
+                            device.major(),
+                            device.minor(),
+                            device.rate()
+                        ),
+                    )?;
+                }
+            }
+
+            if let Some(throttle_write_bps_device) = blkio.throttle_read_iops_device() {
+                for device in throttle_write_bps_device {
+                    let path = self.get_file(&format!("io.max"));
+                    ensure_write_file(
+                        path,
+                        &format!(
+                            "{}:{} riops={}",
+                            device.major(),
+                            device.minor(),
+                            device.rate()
+                        ),
+                    )?;
+                }
+            }
+
+            if let Some(throttle_write_bps_device) = blkio.throttle_write_iops_device() {
+                for device in throttle_write_bps_device {
+                    let path = self.get_file(&format!("io.max"));
+                    ensure_write_file(
+                        path,
+                        &format!(
+                            "{}:{} wiops={}",
+                            device.major(),
+                            device.minor(),
+                            device.rate()
+                        ),
+                    )?;
+                }
+            }
+        }
+
+        if let Some(unified) = res.unified() {
+            for (k, v) in unified {
+                ensure_write_file(self.get_file(k), v)?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+pub fn new(name: String) -> Result<Box<dyn Cgroup>> {
+    // fast path
+    let v1 = PathBuf::from("/sys/fs/cgroup");
+    let v2 = PathBuf::from("/sys/fs/cgroup/unified");
+    if v1.join("cpu").exists() {
+        let stat = statfs(v1.join("cpu").as_path()).map_err(|e| std::io::Error::from(e))?;
+        let is_v1 = stat.filesystem_type() == nix::sys::statfs::CGROUP_SUPER_MAGIC;
+        if is_v1 {
+            if v2.exists() {
+                // Check if we are running in a hybrid cgroup setup
+                let data = std::fs::read(v2.join("cgroup.controllers"))
+                    .map_err(|e| std::io::Error::from(e))?;
+
+                let trimmed = std::str::from_utf8(&data)
+                    .map_err(|e| {
+                        Error::Others(format!(
+                            "could not convert cgroup.controllers to string: {}",
+                            e
+                        ))
+                    })?
+                    .trim();
+
+                if trimmed.len() > 0 {
+                    return Err(Error::FailedPrecondition(
+                        "hybyrid cgroup is not supported".to_string(),
```

Comment:
just for my own understanding: why is hybrid cgroup not supported?

Reply:
Simplicity. It could be added but for now it's either cgroupv1 or cgroupv2.